tools/fee-service/engine/rebalancing.py

```
@@ -0,0 +1,179 @@
+        rebal_cfg = config.get("rebalancing", {})
+        self.fallback_bps = rebal_cfg.get("fallback_default_bps", 1.5)
+        self.lookback_days = rebal_cfg.get("lookback_days", 30)
+        self.route_overrides = rebal_cfg.get("route_overrides", {})
+
+        # Computed from on-chain data
+        self._route_costs: dict[str, dict] = {}
+        # Live data from rebalancer API (future)
+        self._live_costs: dict[str, dict] = {}
+
+    def update_from_tx_data(self, rebalancing_data: dict,
+                            volume_data: dict = None):
```

Ruff (0.15.5)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)